### PR TITLE
api_tracer: fix tracer GC which accidentally pruned the metaroot

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes

- This PR is derived from https://github.com/ethereum/go-ethereum/pull/17357.

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- https://github.com/ethereum/go-ethereum/issues/17355

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: storage/statedb/database.go [hunk 1]
```diff
@@ -677,6 +677,12 @@ func (db *Database) Dereference(root common.Hash) {
 	db.gcLock.Lock()
 	defer db.gcLock.Unlock()
 
+	// Sanity check to ensure that the meta-root is not removed
+	if root == (common.Hash{}) {
```
Comment (ehnuje) [inline storage/statedb/database.go:681]: How about using `common.EmptyHash`?

File: node/cn/api_tracer.go [hunk 2]
```diff
@@ -712,7 +713,6 @@ func (api *PrivateDebugAPI) computeStateDB(block *types.Block, reexec uint64) (*
 		if err := statedb.Reset(root); err != nil {
 			return nil, fmt.Errorf("state reset after block %d failed: %v", block.NumberU64(), err)
 		}
-		database.TrieDB().Reference(root, common.Hash{})
```
Comment (ehnuje) [inline node/cn/api_tracer.go:715]: Why did you remove this line?

Comment (yoomee1313) [reply node/cn/api_tracer.go]: My Mistake.. I recovered it!

File: storage/statedb/database.go (hunk 1 above)
Comment (yoomee1313) [reply storage/statedb/database.go]: Good Idea. I applied at https://github.com/klaytn/klaytn/pull/761/commits/fe9b934ddab35b1c12d6394e83440aa983aa338d.

**Review by KimKyungup — APPROVED** [review]

**Review by ehnuje — DISMISSED** [review]

File: node/cn/api_tracer.go [hunk 3]
```diff
@@ -553,8 +555,14 @@ func (api *PrivateDebugAPI) traceBlock(ctx context.Context, block *types.Block,
...
-	if config != nil && config.TxHash != (common.Hash{}) {
-		if !containsTx(block, config.TxHash) {
+	if config != nil && !common.EmptyHash(config.TxHash) {
+		var exists bool
+		for _, tx := range block.Transactions() {
+			if exists = (tx.Hash() == config.TxHash); exists {
+				break
+			}
+		}
+		if !exists {
```
Comment (aidan-kwon) [inline node/cn/api_tracer.go:565]: We just changed this code from https://github.com/klaytn/klaytn/commit/05749a25def8159d697dc6e5d0158c3c45638e8c
@yoomee1313 @ehnuje Which one is the latest one?

Comment (yoomee1313) [reply node/cn/api_tracer.go]: @aidan-kwon @ehnuje's commit is latest. I removed it https://github.com/klaytn/klaytn/pull/761/commits/d1ab996946a763bf85c922a9f5436a47c656bf1c.

**Review by aidan-kwon — APPROVED** [review]

File: storage/statedb/database.go [hunk 4]
```diff
@@ -677,6 +677,12 @@ func (db *Database) Dereference(root common.Hash) {
 	db.gcLock.Lock()
 	defer db.gcLock.Unlock()
 
+	// Sanity check to ensure that the meta-root is not removed
```
Comment (KimKyungup) [inline storage/statedb/database.go:680]: This check can be moved before 677 line.

Comment (yoomee1313) [reply storage/statedb/database.go:677]: @KimKyungup Moved those lines! https://github.com/klaytn/klaytn/pull/761/commits/ef40dc8d10ed209370c00d4f268cf76aba257b6a

**Review by ehnuje — APPROVED** [review]

**Review by KimKyungup — APPROVED** [review]

**Review by aidan-kwon — APPROVED** [review]

**Review by jeongkyun-oh — APPROVED** [review]
